source/server/admin/stats_render.cc

```diff
@@ -0,0 +1,304 @@
+      histogram.cumulativeStatistics().computedQuantiles();
+  size_t min_size = std::min({computed_quantiles.size(), cumulative_quantiles.size(),
+                              interval_statistics.supportedQuantiles().size()});
+  ASSERT(min_size == computed_quantiles.size());
+  ASSERT(min_size == cumulative_quantiles.size());
+
+  for (size_t i = 0; i < min_size; ++i) {
+    ProtobufWkt::Struct computed_quantile_value;
+    ProtoMap& computed_quantile_value_fields = *computed_quantile_value.mutable_fields();
+    const auto& interval = computed_quantiles[i];
+    const auto& cumulative = cumulative_quantiles[i];
+    computed_quantile_value_fields["interval"] =
+        std::isnan(interval) ? ValueUtil::nullValue() : ValueUtil::numberValue(interval);
+    computed_quantile_value_fields["cumulative"] =
+        std::isnan(cumulative) ? ValueUtil::nullValue() : ValueUtil::numberValue(cumulative);
+
+    computed_quantile_value_array.push_back(ValueUtil::structValue(computed_quantile_value));
+  }
+  computed_quantile_fields["values"] = ValueUtil::listValue(computed_quantile_value_array);
+  histogram_array_.push_back(ValueUtil::structValue(computed_quantile));
+}
+
+// Collects the buckets from the specified histogram, using either the cumulative
+// or disjoint views, as controlled by buckets_fn.
+void StatsJsonRender::collectBuckets(const std::string& name,
+                                     const Stats::ParentHistogram& histogram,
+                                     const UInt64Vec& interval_buckets,
+                                     const UInt64Vec& cumulative_buckets) {
+  const Stats::HistogramStatistics& interval_statistics = histogram.intervalStatistics();
+  Stats::ConstSupportedBuckets& supported_buckets = interval_statistics.supportedBuckets();
+
+  // Make sure all vectors are the same size.
+  ASSERT(interval_buckets.size() == cumulative_buckets.size());
+  ASSERT(cumulative_buckets.size() == supported_buckets.size());
+  size_t min_size =
+      std::min({interval_buckets.size(), cumulative_buckets.size(), supported_buckets.size()});
+
+  ProtobufWkt::Struct histogram_obj;
+  ProtoMap& histogram_obj_fields = *histogram_obj.mutable_fields();
+  histogram_obj_fields["name"] = ValueUtil::stringValue(name);
+
+  std::vector<ProtobufWkt::Value> bucket_array;
+  for (size_t i = 0; i < min_size; ++i) {
+    ProtobufWkt::Struct bucket;
+    ProtoMap& bucket_fields = *bucket.mutable_fields();
+    bucket_fields["upper_bound"] = ValueUtil::numberValue(supported_buckets[i]);
+
+    // ValueUtil::numberValue does unnecessary conversions from uint64_t values to doubles.
+    bucket_fields["interval"] = ValueUtil::numberValue(interval_buckets[i]);
+    bucket_fields["cumulative"] = ValueUtil::numberValue(cumulative_buckets[i]);
+    bucket_array.push_back(ValueUtil::structValue(bucket));
+  }
+  histogram_obj_fields["buckets"] = ValueUtil::listValue(bucket_array);
+  histogram_array_.push_back(ValueUtil::structValue(histogram_obj));
```

Comment:
May be we can use protobuf list here directly as the `bucket_array` to avoid uncessary struct value copies?

Reply:
done